File: plugins/distortion/ring-mod/ring-mod.js
// Ring Modulator — multiplies the input by a sine carrier. Output spectrum
// is the sum-and-difference of every input partial with the carrier, so
// harmonic content scrambles into inharmonic sidebands. At low carrier
// frequencies this is a tremolo; in the speech-band it produces the
// metallic Dalek voice; above 1 kHz it dissolves the source into bell-like
// inharmonic textures.

function RingModProcessor() {
  this.phase = 0;
}
RingModProcessor.parameterDescriptors = [
  { name: 'frequency', label: 'Carrier (Hz)', defaultValue: 220, minValue: 1, maxValue: 2000 },
  { name: 'mix',       label: 'Mix',          defaultValue: 1,   minValue: 0, maxValue: 1 }
];
RingModProcessor.prototype.process = function(inputs, outputs, parameters) {
  var input = inputs[0][0], output = outputs[0][0];
  if (!input || !output) return true;
  var freq = parameters.frequency[0];
  var mix = parameters.mix[0];
  var phaseInc = (2 * Math.PI * freq) / sampleRate;
  for (var i = 0; i < input.length; i++) {
    var mod = Math.sin(this.phase);
    var wet = input[i] * mod;
    output[i] = input[i] * (1 - mix) + wet * mix;
    this.phase += phaseInc;
    if (this.phase > 2 * Math.PI) this.phase -= 2 * Math.PI;
  }
  return true;
};

registerProcessor('ring-mod', RingModProcessor);

File: plugins/distortion/ring-mod/RingMod.java
package com.vocalmonitor.plugin.community;

import com.vocalmonitor.plugin.VocalMonitorNativePlugin;

// Ring Mod (native port) — multiplies input by a sine carrier. Low
// carrier = tremolo; speech-band = Dalek robot voice; >1 kHz = bell-like
// clangorous inharmonic texture.
public final class RingMod implements VocalMonitorNativePlugin {
    private float phase = 0f;
    private int sampleRate = 44100;
    private float frequency = 220f, mix = 1f;

    @Override public void init(int sr) { this.sampleRate = sr; phase = 0f; }
    @Override public String[] parameterNames() { return new String[] { "frequency", "mix" }; }
    @Override public float parameterMin(String n) { return "frequency".equals(n) ? 1f : 0f; }
    @Override public float parameterMax(String n) { return "frequency".equals(n) ? 2000f : 1f; }
    @Override public float parameterDefault(String n) { return "frequency".equals(n) ? 220f : 1f; }
    @Override public String parameterLabel(String n) { return "frequency".equals(n) ? "Carrier (Hz)" : "Mix"; }
    @Override public void setParameter(String n, float v) {
        if ("frequency".equals(n)) frequency = v;
        else if ("mix".equals(n)) mix = v;
    }

    @Override
    public void process(float[] input, float[] output) {
        final float phaseInc = (float) (2.0 * Math.PI * frequency / sampleRate);
        final float twoPi = (float) (2.0 * Math.PI);
        final float mixLocal = mix;
        final float dry = 1f - mixLocal;
        float ph = phase;
        final int n = input.length;
        for (int i = 0; i < n; i++) {
            float mod = (float) Math.sin(ph);
            float wet = input[i] * mod;
            output[i] = input[i] * dry + wet * mixLocal;
            ph += phaseInc;
            if (ph > twoPi) ph -= twoPi;
        }
        phase = ph;
    }
}
